zkevm-circuits/src/sha256_circuit.rs

```diff
@@ -123,8 +126,27 @@ impl SubCircuit<Fr> for SHA256Circuit<Fr> {
 
         for hash_event in &self.0 {
             hasher.update(layouter, chng, &hash_event.input)?;
-            // TODO: verify output and digest in event
-            let _ = hasher.finalize(layouter, chng)?;
+
+            let digest = hasher.finalize(layouter, chng)?;
+            let ref_digest = hash_event
+                .digest
+                .chunks_exact(4)
+                .map(|bt| bt.iter().fold(0u32, |sum, v| sum * 256 + *v as u32))
+                .collect::<Vec<_>>();
+            for (w, check) in digest.into_iter().zip(ref_digest) {
+                w.0.assert_if_known(|digest_word| *digest_word == check);
+            }
+
+            if hasher.blocks() > self.1 {
+                log::error!("handled 512-bit block exceed limit ({})", self.1);
+                return Err(Error::Synthesis);
+            }
+        }
+
+        // paddings
+        for _i in hasher.blocks()..self.1 {
+            hasher.update(layouter, chng, &[])?;
+            hasher.finalize(layouter, chng)?;
```

Comment:
will the padding be slow?  I remembered previously, we re-calculate the witness for keccak("") repeatedly.  Quite slow.  Later we first calculate the witness values only once, and then just write them into cells.  Much faster.

Reply:
Have made a quick bench with `k=19`, by calculating only 1x 512-bit block and 128x blocks, the total proving time is:

+ 128x blocks: 17.355s
+ 1x block: 16.793s

So the overhead for synthising is ~3%, i think it is trivial so such performance enhancement is not critical.

I guess the reasons may be:

1. The number of sha256 is small and "data density" is sparse (it need 2k rows for a 512-bit block)
2. Our circuit do not need extra calculation for the input bytes.